### ISSUE TITLE
Sharing python code between recipes

### ISSUE THREAD

**Issue by memsharded** [conan-io/conan#474]

**Comment by annulen** [comment]
I think the best way to achieve this is to have special conan packages that install python modules which can be imported in conanfile.py

**Comment by memsharded** [comment]
I have done some analysis of this feature, and it is very unlikely to be possible to be implemented in conan, because of the way things are processed:

1. Starting from the user conanfile.txt or conanfile reference, the system retrieves such recipe
2. Then it evaluates such recipe, taking into account the settings, and options, processing the configuration, etc, to decide which dependencies/requirements it needs. Such requirements can be even conditional
3. So conan needs to evaluate such recipe, in a python evaluation, and if such python recipe has ``imports`` to other python code, and such python code leaves in another package, that package would be up in the dependency graph, and as such, has not been retrieved yet.

There could be a restricted way to do so, if the ``imports`` of such common python code would be restricted to the inside of the ``source()``, ``build()`` and ``package()`` methods, which are the ones executed after the dependency tree is defined. Still have to figure out how to do this.

If your idea is to decouple it from the dependency tree, it might be easier. Something like:
1. Have a recipe for "tools", like MinGW, CMake, or "MyCommonPythonCode".
2. Use the conan virtualenvs generator, run such recipe to define your build environment, and that would include putting some python code in the PYTHONPATH
3. Now run conan as usual. It is not necessary that packages define a dependency to CMake or the common python code, they will be defined in the environment paths

The inconvenience for this simpler approach is that if you don't set up the environment tools before, then it will fail when you try to install the project dependencies.

**Comment by annulen** [comment]
>There could be a restricted way to do so, if the imports of such common python code would be restricted to the inside of the source(), build() and package() methods, which are the ones executed after the dependency tree is defined.

I think this restriction is unncessary, you need just to replace import statements with your own code by preprocessing conanfile, and do actual imports after dependency tree is known. Relying on PYTHONPATH looks like a bad idea IMO.

**Comment by memsharded** [comment]
But I mean. Lets suppose you have a common python function that helps you decide if you are on Windows (dummy example, to make it simple):

```python
def is_windows(settings):
    return settings.os == "Windows" 
```
Now you decide this is a common function, you want to use on all your package recipes. So you create a conan package ``MyPyFunctions/0.1@myuser/stable``.

The packages that want to use that function, will add a ``requires="MyPyFunctions/0.1@myuser/stable"``, and use the function, for example:

```python
import my_python_functions
...
def configure(self):
    if is_windows(self.settings):
         self.requires("SomWinPkg/1.2@user/channel")
```

This code cannot be preprocessed. It has to be fully evaluated and executed before we really know the dependencies of the current package and can retrieve those dependencies. You will not have ``MyPyFunctions/0.1@myuser/stable`` in your hard drive and then ``is_windows()`` won't exist and won't be possible to use. But conan needs to execute the ``configure()`` method to keep evaluating the dependency tree. So it is a chicken and egg problem, that can only be solved by confining the imports and dependent code to such functions that are evaluated after the construction of the dependency tree.

**Comment by annulen** [comment]
Yep, configure code surely cannot use external modules in this approach

**Comment by memsharded** [comment]
Exactly, that was the point, that is why I said, that such reuse of code should be just confined to those ``build, source, package`` methods, otherwise, it would fail. Seems still a bit artificial, trying to figure out some other way to achieve the same goal, without much success.

Probably, a way could be to slightly abuse ``exports``, pointing it to your folder when the common python code lives, so it will be copied into the recipe. Yes, it is duplicated code, but maybe it is better, being explicit. What I don't like about importing python code is that you can actually be changing the recipe completely without knowing, because you have changed the internal code by importing. Copying the code creates a complete snapshot for the recipe. Just an idea.

**Comment by annulen** [comment]
For example, in Gentoo ebuilds share a lot of "library" code which is a part of portage tree, not `portage` application. This allows them to evolve separately, decreases burden on `portage` application developers, and allows packagers to create ebuild classes which simplify ebuild code for packages using certain build systems, e.g. by providing helpers to map use-flags (options in Conan parlance) to `configure` flags and so on.

Explicit duplication of code will require a lot of discipline to avoid conciousless diverging, and it won't give newcomers easy idea that it's possible to copy necessary stuff from elsewhere instead of reinventing the wheel.

**Comment by memsharded** [comment]
Released in conan 0.13.1
